### ISSUE TITLE
simp diverges in presence of match

### ISSUE THREAD

**Issue by gert-smolka** [mattam82/Coq-Equations#171]
simp diverges, see example below (Coq 8.9).
```
From Coq Require Import Arith Lia.
From Equations Require Import Equations.

Equations? Mod (x y : nat) : nat by wf x lt :=
 { Mod x 0 := 0 ;
   Mod x y := if le_lt_dec y x then Mod (x - y) y else x } .
Proof.
 lia.
Qed.

Fact Mod_lt x y :
  Mod x (S y) < S y.
Proof.
  induction x as [x IH] using lt_wf_ind.
  simp Mod.
Abort.

```

**Comment by mattam82** [comment]
Yes, “simp” generates an unconditional equation for the second clause while rewriting with it should really depend on the conditional “le_lt_dec y x”. If you use a “with” clause instead then the rewriting will be properly guarded by the test. In general, equations of well-founded definitions do not generate a terminating rewrite system!

**Comment by mattam82** [comment]
One way to fix this would be to generate conditional equations in that case

**Comment by gert-smolka** [comment]
Maybe disallow native matches in Equations.

**Comment by mattam82** [comment]
I could at least warn if there is a recursive call under a match in an equation, it’s an easy syntactic check.

**Comment by TDiazT** [comment]
Not entirely sure if related or not, but I got a stack overflow while trying to prove an obligation for a wf definition:

Goal : 
```
queries_size φ < query_size (NestedField f α φ) + queries_size queries
```

with 
```
Equations query_size query : nat :=
    {
      query_size (NestedField _ _ q') := 1 + queries_size q';
      query_size (NestedLabeledField _ _ _ q') => 1 + (queries_size q');
      query_size (InlineFragment _ q') => 1 + (queries_size q');
      query_size _ := 1
    }
  where
  queries_size queries : nat :=
    {
      queries_size [::] := 0;
      queries_size (cons hd tl) := query_size hd + queries_size tl
    }.
```

If I use `simp query_size` I get the SO.


(Maybe using equations for that function wasn't necessary haha, but still)

**Comment by mattam82** [comment]
@Tdiazt that looks unrelated. It’s probably autorewrite looping for another reason. Can you try rewriting with the equations (queries_size_equation_1 and so on) to understand how it could loop?

**Comment by TDiazT** [comment]
I redefined it so I no longer have this situation but I recall that rewriting with the equations would just give me the expected result. In this case: `1 + queries_size q'`. 

I think I may have another example... I'll post it as soon as I find it again.

**Comment by mattam82** [comment]
It might be because `simp f` calls also `typeclasses eauto with f Below subterm_relation rec_decision`, where `f` represents the hint database associated to `f`. (Print HintDb f to see its contents).
